hpx/util/detail/unwrap_impl.hpp

```diff
@@ -0,0 +1,294 @@
+//  Copyright (c) 2017 Denis Blank
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_UTIL_DETAIL_UNWRAP_IMPL_HPP
+#define HPX_UTIL_DETAIL_UNWRAP_IMPL_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/traits/future_traits.hpp>
+#include <hpx/traits/is_future.hpp>
+#include <hpx/traits/is_tuple_like.hpp>
+#include <hpx/util/invoke.hpp>
+#include <hpx/util/invoke_fused.hpp>
+#include <hpx/util/pack_traversal.hpp>
+#include <hpx/util/result_of.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace hpx {
+namespace util {
+    namespace detail {
+        /// This struct defines a configurateable environment in order
+        /// to adapt the unwrapping behaviour.
+        ///
+        /// The old unwrap (unwrapped) allowed futures to be instanted
+        /// with void whereas the new one doesn't.
+        ///
+        /// The old unwrap (unwrapped) also unwraps the first occuring future
+        /// independently a single value or multiple ones were passed to it.
+        /// This behaviour is inconsistent and was removed in the
+        /// new implementation.
+        ///
+        /// \note If the old implementation of unwrapped gets removed all
+        ///       the configuration mechanis here can be removed while
+        ///       preserving the behaviour of the new implementations
+        ///       configuration.
+        ///
+        template <bool AllowsVoidFutures, bool UnwrapTopLevelTuples>
+        struct unwrap_config
```

Comment:
You might want to introduce a preprocessor compatibility constant right away and use it to wrap this code.

Our policy for introducing breaking changes is to do it in 3 stages (releases):

1. Wrap the old API in a preprocessor macro, but leave it ON by default. Also mark the old API as deprecated (at least in the docs)
2. Change the macro to be OFF by default, allowing for the user to still revert to the old API 
3. Remove the old API altogether